### PR TITLE
Convert Expire command to c

### PR DESCRIPTION
**Pull request by cmwshang**
Note: it will not be implemented fully until remote and s3 storage functions are completed.

### CODE REVIEW

**Comment by dwsteele** [comment]
The expire test fixes have been committed at 19d8358cba00 and linefeed changes in `backupCommonTest.c` at 2b75f57232e3d2c983e0a703.

File: src/info/infoBackup.c [hunk 1]
```diff
@@ -311,6 +313,64 @@ infoBackupData(const InfoBackup *this, unsigned int backupDataIdx)
...
+    // Return a 0 sized list if no current backups or none matching the filter
+    StringList *result = strLstNew();
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // For each backup label, compare it to the filter (if any) and sort it for return
+        for (unsigned int backupLabelIdx = 0; backupLabelIdx < infoBackupDataTotal(this); backupLabelIdx++)
+        {
+            InfoBackupData backupData = infoBackupData(this, backupLabelIdx);
+            if (param.filter == NULL || regExpMatchOne(param.filter, backupData.backupLabel))
```
Comment (dwsteele) [inline src/info/infoBackup.c]: It's a bit wasteful to use regExpMatchOne here.  Parsing regexps is relatively expensive.

Create the RegExp in advance and use it multiple times like storagePosixList does.

File: src/info/infoBackup.c [hunk 2]
```diff
@@ -311,6 +313,64 @@ infoBackupData(const InfoBackup *this, unsigned int backupDataIdx)
...
+
+    // Return a 0 sized list if no current backups or none matching the filter
+    StringList *result = strLstNew();
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // For each backup label, compare it to the filter (if any) and sort it for return
+        for (unsigned int backupLabelIdx = 0; backupLabelIdx < infoBackupDataTotal(this); backupLabelIdx++)
+        {
+            InfoBackupData backupData = infoBackupData(this, backupLabelIdx);
```
Comment (dwsteele) [inline src/info/infoBackup.c]: Add a linefeed here.

File: src/info/infoBackup.c [hunk 3]
```diff
@@ -311,6 +313,64 @@ infoBackupData(const InfoBackup *this, unsigned int backupDataIdx)
...
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(INFO_BACKUP, this);
+        FUNCTION_LOG_PARAM(STRING, backupDeleteLabel);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+
+    for (unsigned int idx = 0; idx < infoBackupDataTotal(this); idx++)
+    {
+        InfoBackupData backupData = infoBackupData(this, idx);
```
Comment (dwsteele) [inline src/info/infoBackup.c]: Add linefeed here.

File: src/info/infoBackup.c [hunk 4]
```diff
@@ -311,6 +313,64 @@ infoBackupData(const InfoBackup *this, unsigned int backupDataIdx)
...
+        if (strCmp(backupData.backupLabel, backupDeleteLabel) == 0)
+        {
```
Comment (dwsteele) [inline src/info/infoBackup.c]: No braces needed on this if.

File: src/command/expire/expire.c [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+    FUNCTION_LOG_END();
+
+    ASSERT(infoBackup != NULL);
+
+    unsigned int result = 0;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // ??? Retention options will need to be indexed
+        unsigned int differentialRetention = cfgOptionTest(cfgOptRepoRetentionDiff) ?
```
Comment (dwsteele) [inline src/command/expire/expire.c]: Why is ??? still here?

File: src/command/expire/expire.c [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+    FUNCTION_LOG_END();
+
+    ASSERT(infoBackup != NULL);
+
+    unsigned int result = 0;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // ??? Retention options will need to be indexed
+        unsigned int fullRetention = cfgOptionTest(cfgOptRepoRetentionFull) ?
```
Comment (dwsteele) [inline src/command/expire/expire.c]: Why is ??? still here?

File: src/command/expire/expire.c [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+
+        // Find all the expired differential backups
+        if (differentialRetention > 0)
+        {
+            // Get a list of full and differential backups. Full are considered differential for the purpose of retention.
+            // Example: F1, D1, D2, F2, repo-retention-diff=2, then F1,D2,F2 will be retained, not D2 and D1 as might be expected.
+            StringList *currentBackupList = infoBackupDataLabelListP(
+                infoBackup, .filter = backupRegExpP(.full = true, .differential = true));
+
+            // If there are more backups then the number to retain, then expire the oldest ones
```
Comment (dwsteele) [inline src/command/expire/expire.c]: "than the number"

File: src/command/expire/expire.c [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+    }
+    MEM_CONTEXT_TEMP_END();
+
+    FUNCTION_LOG_RETURN(UINT, result);
+}
+
+/***********************************************************************************************************************************
+Log detailed information about archive logs removed
+***********************************************************************************************************************************/
+void
```
Comment (dwsteele) [inline src/command/expire/expire.c]: This function should be static.

File: src/command/expire/expire.c [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+                                    storageRepo(), strNewFmt(STORAGE_REPO_ARCHIVE "/%s", strPtr(archiveId)));
+                                storagePathRemoveP(storageRepoWrite(), fullPath, .recurse = true);
+                                LOG_INFO("remove archive path: %s", strPtr(fullPath));
+                            }
+
+                            // Continue to next directory
+                            break;
+                        }
+
+                        // If get here, then a local backup was found for retention
```
Comment (dwsteele) [inline src/command/expire/expire.c]: If we get here?

File: src/command/expire/expire.c [hunk 10]
```diff
@@ -0,0 +1,680 @@
...
+                                        strNewFmt(STORAGE_REPO_ARCHIVE "/%s", strPtr(archiveId)), .errorOnMissing = true,
+                                        .expression = STRDEF(WAL_SEGMENT_DIR_REGEXP)),
+                                    sortOrderAsc);
+
+                            for (unsigned int walIdx = 0; walIdx < strLstSize(walPathList); walIdx++)
+                            {
+                                String *walPath = strLstGet(walPathList, walIdx);
+                                removeArchive = true;
+
+                                LOG_DEBUG("Expire process, found major WAL path: %s", strPtr(walPath));
```
Comment (dwsteele) [inline src/command/expire/expire.c]: We already have log entries to tell us we are in the expire process so need to repeat that here.  Debug level logging automatically prints the module/function as well.

Ditto for other places where this is output.

File: src/command/expire/expire.c [hunk 11]
```diff
@@ -0,0 +1,680 @@
...
+                                // Remove the entire directory if all archive is expired
+                                if (removeArchive)
+                                {
+                                    String *fullPath = strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strPtr(archiveId), strPtr(walPath));
+                                    storagePathRemoveP(storageRepoWrite(), fullPath, .recurse = true);
+                                    LOG_DEBUG("Expire process, remove major WAL path: %s", strPtr(fullPath));
+                                    archiveExpire.total++;
+                                    archiveExpire.start = strDup(walPath);
+                                    archiveExpire.stop = strDup(walPath);
+
```
Comment (dwsteele) [inline src/command/expire/expire.c]: Remove linefeed.

File: src/command/expire/expire.c [hunk 12]
```diff
@@ -0,0 +1,680 @@
...
+                            // If this is the backup selected for retention, store its data
+                            if (strCmp(archiveIdBackup.backupLabel, strLstGet(localBackupArchiveRententionList, 0)) == 0)
+                                archiveRetentionBackup = infoBackupData(infoBackup, infoBackupIdx);
+
+                            // If this is a backup associated with this archive Id, then add it to the list to check
+                            if (archiveIdBackup.backupPgId == archivePgId)
+                                lstAdd(archiveIdBackupList, &archiveIdBackup);
+                        }
+
+                        bool removeArchive = false;
```
Comment (dwsteele) [inline src/command/expire/expire.c]: Put this bool between the comment and the if.

File: src/info/infoBackup.h [hunk 13]
```diff
@@ -61,6 +62,25 @@ Functions
...
+{
+    const String *filter;
+} InfoBackupDataLabelListParam;
+
+#define infoBackupDataLabelListP(this, ...)                                                                                        \
+    infoBackupDataLabelList(this, (InfoBackupDataLabelListParam){__VA_ARGS__})
+#define infoBackupDataLabelListNP(this)                                                                                            \
+    infoBackupDataLabelList(this, (InfoBackupDataLabelListParam){0})
+
+StringList *infoBackupDataLabelList(const InfoBackup *this, InfoBackupDataLabelListParam param);
```
Comment (dwsteele) [inline src/info/infoBackup.h]: As we discussed, have P/NP variants for a function with one optional param seems to be overkill.

File: src/command/expire/expire.c (hunk 5 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: Was just a note to remind us to do it. Removed.

File: src/command/expire/expire.c (hunk 7 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: corrected

File: src/command/expire/expire.c (hunk 6 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: Just a note to remind us. Will remove.

File: src/command/expire/expire.c (hunk 8 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: done.

File: src/command/expire/expire.c (hunk 9 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: We never use pronouns in the comments which is why it is omitted.

File: src/command/expire/expire.c (hunk 12 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: done

File: src/command/expire/expire.c (hunk 10 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: Removed "Expire process, " from all logs. However, I believe we no longer have them in the expect logs, so I would like to suggest I add testing for them in the unit test.

File: src/command/expire/expire.c (hunk 11 above)
Comment (cmwshang) [reply src/command/expire/expire.c]: done

File: src/info/infoBackup.c (hunk 3 above)
Comment (cmwshang) [reply src/info/infoBackup.c]: done

File: src/info/infoBackup.c (hunk 4 above)
Comment (cmwshang) [reply src/info/infoBackup.c]: removed

File: src/info/infoBackup.c (hunk 2 above)
Comment (cmwshang) [reply src/info/infoBackup.c]: done

File: src/info/infoBackup.c (hunk 1 above)
Comment (cmwshang) [reply src/info/infoBackup.c]: done

File: src/info/infoBackup.h (hunk 13 above)
Comment (cmwshang) [reply src/info/infoBackup.h]: removed - now pass "this" and "expression"

File: test/src/module/info/infoBackupTest.c [hunk 14]
```diff
@@ -192,6 +194,35 @@ testRun(void)
...
+            strPtr(strLstJoin(infoBackupDataLabelListP(infoBackup, .filter = backupRegExpP(.full=true)), ", ")),
+            "20161219-212741F", "  full=true");
+        TEST_RESULT_STR(
+            strPtr(strLstJoin(infoBackupDataLabelListP(infoBackup, .filter = backupRegExpP(.differential=true)), ", ")),
+            "20161219-212741F_20161219-212803D", "differential=true");
+        TEST_RESULT_STR(
+            strPtr(strLstJoin(infoBackupDataLabelListP(infoBackup, .filter = backupRegExpP(.incremental=true)), ", ")),
+            "20161219-212741F_20161219-212918I", "incremental=true");
+
+
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply test/src/module/info/infoBackupTest.c]: done

**Comment by dwsteele** [comment]
Committed all info/* modules changes at c64c9c05905d.

**Comment by dwsteele** [comment]
Committed at 434cd832855e4e18.

**Pull request closed without merging**